### PR TITLE
feat: Icarus mod support — Firestore source + .exmodz PAK compilation (#136)

### LINKED ISSUE #136: EPIC: Icarus mod support — .exmod/.exmodz PAK compilation on Linux
Icarus is Unreal-Engine-based and ships two mod formats: ready-made `.pak` files (already deployable via the existing source/linker pipeline once a source exists), and `.exmodz` — a zip bundling a JSON diff (`.EXMOD`) against the base game's data tables plus optional pre-built `.uasset`/`.uexp` files. Turning `.exmodz` into a loadable `.pak` on Linux (no Unreal Editor, no Wine) is the actual scope of this work — see `IDEAS.md`.

A research spike + design + implementation plan already exist, not yet merged, on branch `docs/icarus-exmod-pak-research`:

- Design: `docs/plans/2026-07-29-icarus-exmod-pak-research.md`
- Implementation plan: `docs/superpowers/plans/2026-07-29-icarus-exmod-pak-compilation.md`

## Key findings (grounded against a real sample `.EXMODZ`, `icarus-mod-tools`/`project_daedalus`, and public modding-community sources)

- Icarus runs UE 4.26.2/4.27, classic (non-IoStore) `.pak` format, **not encrypted** — community routinely round-trips `data.pak` with plain `UnrealPak.exe`, no AES-key step anywhere.
- `.EXMOD` diffs target **plain JSON data tables** already shipped inside the game's pak (row-keyed field overrides), not compiled binary assets — no Unreal "cook" step needed.
- [`repak`](https://github.com/trumank/repak) proves this format is readable/writable natively on Linux; no mature Go pak library exists, so this plan hand-rolls a small `internal/unrealpak` reader/writer scoped to uncompressed/unencrypted only.
- Mod catalog metadata lives in Firestore with **no existing public JSON API** — plan is to read Firestore's own public REST API directly (unauthenticated, per design doc decision), needing a hand-written Go `ModSource` rather than a declarative YAML custom source.

## Scope

- New `internal/unrealpak` package: PAK footer/index reader, file reader, writer, round-trip tested.
- New `internal/source/icarus` package: Firestore-backed `ModSource`, `.EXMOD` diff parsing + row-patch application, `.EXMODZ` unpacking, compile orchestration.
- New `domain.DeployCompile` mode + `source.Compiler` optional interface wiring the compile step into `Service`'s existing cache-population path (no linker changes needed).
- Opens with an empirical PAK-format validation spike against a real local install — the design's stated biggest open risk.

## Status

Design and plan are written and self-reviewed; **implementation has not started**. Multi-week effort per the design doc's own estimate — no existing library shortcuts the hard part.

## Done when

`.exmodz` mods install end-to-end through LMM (search/download via the Icarus source → compiled `_P.pak` → deployed), validated against a real local Icarus install.

### PR DESCRIPTION
**Pull request by dyoung522**
## Summary

Implements EPIC #136: full Icarus mod support — browsing/installing mods from the community Firestore catalog, and compiling `.exmodz` mods into engine-loadable `_P.pak` files natively on Linux (no Unreal Editor, no Wine).

- **`internal/unrealpak`** (game-agnostic, zero lmm imports): UE v11 PAK reader + writer — 221-byte footer, primary index with bit-packed encoded entries, path-hash index (FNV-1a-64, seeded), full directory index, 53-byte local headers. Every structural constant was empirically verified against the real game's 34 paks / 173,078 entries before implementation; the reader opens the shipping game's files, and the writer's output is byte-round-trip-verified.
- **`internal/source/icarus`**: Firestore REST catalog client (public read-only, project `projectdaedalus-fb09f`, live-verified against 538 real mods), registered as a built-in source beside NexusMods/CurseForge; `.EXMOD` diff engine (hyphen→slash mount-path resolution, `EndOfMod` sentinel handling — both corrected against a real mod after the planned algorithm resolved 0/14 real rows); `.EXMODZ` unpacking with traversal-safe asset paths; hybrid base-table provider (hosted per-week community dumps + `data_dump_path` local-dir override, byte-compare validation against the install's 40 uncompressed-stored tables).
- **Core wiring**: `domain.DeployCompile`, `source.Compiler` capability, per-`.exmodz`-file compile gating in `Service`'s cache-population path (plain `.pak` mods keep the existing extract/copy path), `data_dump_path` in games.yaml, README/configuration docs, CHANGELOG under `[Unreleased]`. No linker changes; no new dependencies (`go.mod` untouched, stdlib only); CLI and TUI both pick the source up through the shared registry.

## Verification

- Full suite green (19 packages), gofmt/vet clean, trunk check: 0 new findings.
- SDD per-task reviews on all 14 tasks + final whole-branch review (clean after one mechanical lint fix); three empirical spikes recorded in the findings doc on the `docs/icarus-exmod-pak-research` branch.
- Real-data checks: reader enumerates the real `pakchunk0` (9,295 entries) and `data.pak` (298 tables); real `Bear_Mount.EXMODZ` resolves 14/14 rows and 26/26 assets.

## Known limitations (by design, fail-loud)

- A real end-to-end compile is currently blocked by ecosystem state: the hosted dump repo's newest week (236) trails the installed game (week 243), and the pipeline loudly refuses mismatched base data (verified against the real install). It unblocks when the dump repo updates or via a user-supplied `data_dump_path`.
- Oodle-compressed base tables are never decompressed locally (no Go decoder exists); base truth comes from the dumps.
- Follow-up issues to be filed: unknown `deploy_mode`/`link_method` values silently fall back (pre-existing convention, needs a repo-wide decision); `lmm import` local ingest doesn't compile `DeployCompile` files.

Closes #136 (pending post-merge in-game validation).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/unrealpak/reader.go [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+// readRegion reads size bytes at offset and verifies them against want. Every
+// index region in a version-11 pak is SHA1-gated: the footer covers the
+// primary index, and the primary index covers each sub-index. All three gates
+// are enforced — a mismatch is corruption or an unrecognized layout, never
+// something to parse through.
+func readRegion(r io.ReaderAt, offset, size int64, want [20]byte) ([]byte, error) {
+	if offset < 0 || size < 0 {
+		return nil, fmt.Errorf("%w: negative region offset/size", ErrUnsupportedFormat)
+	}
+	buf := make([]byte, size)
```
[the remark this answers is not in the document]
Comment (dyoung522) [reply internal/unrealpak/reader.go]: Fixed in 4d905f5 — `readRegion`/`ReadFile` now validate size fields (non-negative, ≤ file size) via a shared `validateAllocSize` helper before allocating. One correction for the record: `make([]byte, int64)` does compile in Go (make accepts any integer type; the suite was green) — the real issue was the missing bounds validation on untrusted size fields, which is what this addresses.

File: internal/unrealpak/reader.go [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+type cursor struct {
+	b   []byte
+	pos int
+	err error
+}
+
+func (c *cursor) take(n int) []byte {
+	if c.err != nil {
+		return make([]byte, n)
+	}
```
[the remark this answers is not in the document]
Comment (dyoung522) [reply internal/unrealpak/reader.go:388]: Fixed in 4d905f5 — `cursor.take` now clamps n to ≥ 0 in the err-latched early-return path, with a regression test for corrupted negative length fields.

File: internal/source/icarus/firestore_client.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+// and public, with no server-side query support in play — see the design
+// doc's "fetch-all + filter client-side" decision).
+func (c *firestoreClient) listCollection(ctx context.Context, collection string) ([]firestoreDoc, error) {
+	var all []firestoreDoc
+	pageToken := ""
+	for {
+		url := fmt.Sprintf("%s/%s?pageSize=200", c.documentsURL(), collection)
+		if pageToken != "" {
+			url += "&pageToken=" + pageToken
+		}
```
[the remark this answers is not in the document]
Comment (dyoung522) [reply internal/source/icarus/firestore_client.go]: Fixed in 4d905f5 — `nextPageToken` is now `url.QueryEscape`d, with a mock-server round-trip test using a '+'-bearing token.

File: internal/source/icarus/compile.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	// the fail-loud-and-clean contract on this path; the success path
+	// (err == nil here) is untouched.
+	defer func() {
+		if err == nil {
+			return
+		}
+		if rmErr := os.Remove(outputPakPath); rmErr != nil && !os.IsNotExist(rmErr) {
+			err = fmt.Errorf("%w (additionally, removing partial output %s failed: %v)", err, outputPakPath, rmErr)
+		}
+	}()
```
[the remark this answers is not in the document]
Comment (dyoung522) [reply internal/source/icarus/compile.go:75]: Fixed in 4d905f5 — the error-path cleanup defer now closes the Writer (ignoring its close error) before removing the partial output, so the fd can't leak and removal works on platforms that disallow deleting open files.

File: internal/unrealpak/writer.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		offset, size := int64(data.Len()), int64(len(file.data))
+		if offset > math.MaxUint32 || size > math.MaxUint32 {
+			w.f.Close() //nolint:errcheck
+			return fmt.Errorf("unrealpak: %s: offset/size exceeds the 32-bit encoded-entry form this writer emits", file.path)
+		}
+		data.Write(storedEntryHeader(size, file.data))
+		data.Write(file.data)
+
+		locations[file.path] = int32(encoded.Len())
+		// The 12-byte stored record: offset/uncompressed-size/size all
```
[the remark this answers is not in the document]
Comment (dyoung522) [reply internal/unrealpak/writer.go:112]: Fixed in 8b282f6 — Writer.Close now fails loudly (matching the existing >4GiB offset error style) if the encoded-entry blob exceeds math.MaxInt32 before any int32 cast, with a boundary-condition unit test.

File: internal/source/icarus/datadump.go [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+		// The repo also carries a stale "data/" copy of the tree; the
+		// authoritative tables are the root-level ones.
+		if rel == "" || strings.HasPrefix(rel, "data/") {
+			continue
+		}
+		body, err := io.ReadAll(tr)
+		if err != nil {
+			return nil, fmt.Errorf("icarus: reading %s from base-table dump: %w", rel, err)
+		}
+		dump.tables[path.Clean(rel)] = toCRLF(body)
```
[the remark this answers is not in the document]
Comment (dyoung522) [reply internal/source/icarus/datadump.go]: Fixed in 8b282f6 — tar entries are rejected if the declared header size exceeds 64 MiB (largest real table is ~7.3 MB) and reads go through io.LimitReader bounded by the declared size, so neither an oversized nor a lying header can allocate unbounded memory. Oversized-entry fixture test added.

**Comment by dyoung522** [comment]
Also addressed the suppressed comment on `icarus.go:236` in 8b282f6: `fileNameFromURL`'s fallback now returns a properly dotted `mod.<ext>` name, appends the expected extension when a parsed basename has none, and routes degenerate basenames (`..`, `.`, `/`) to the fallback — closing the silent `.exmodz`-routing gap. Table-driven test added.

**Comment by dyoung522** [comment]
Round-3 suppressed comments all addressed in e18b0bd: validateAllocSize now also guards platform-int width; the dead `cacheDir` field is gone and DumpStore's doc comment matches what it does; `ParseExmodz` normalizes backslash entry names and matches the manifest and `.uasset`/`.uexp` patterns case-insensitively, with multiple candidate manifests now a loud error instead of silent last-one-wins. Suite/vet/trunk clean.

**Comment by dyoung522** [comment]
Round-4 suppressed comment addressed in 1c261fe: Search results are now sorted (Name, ID tiebreak — matching the custom source's convention) before pagination, so paging is deterministic. Test pins the ordering.

**Comment by dyoung522** [comment]
Round-5 triage (both suppressed comments):
- `cmd/lmm/root.go` SetDataDir comment: fixed — reworded to what the hook actually does today (gates DumpStore construction; the directory value itself is currently unused, reserved for future dump caching), exactly per the suggestion.
- `CheckUpdates` N+1: acknowledged but deliberately unchanged. For typical installed-mod counts (< ~20) per-mod `GetMod` is cheaper than paging the full ~540-doc catalog, and correctness is unaffected. If profiles grow enough to flip that trade-off, a catalog-map approach belongs with the other source-interaction-quality work tracked in EPIC #127 — noted there rather than optimized speculatively here.
